### DISCUSSION TITLE
Async support

### DISCUSSION THREAD

**Discussion by danny-bos-developer** [ardalis/GuardClauses#226] [Ideas]
Any chance you can also add extension method to make async predicate calls?

I created an extension of my own, it would be nice if all the Guard.Against api's that use a predicate also have an async variant.

```
public static async Task<T> InvalidInputAsync<T>(this IGuardClause guardClause, T input, string parameterName, Func<T, Task<bool>> predicate, string? message = null)
        {
            if (!await predicate(input))
            {
                throw new ArgumentException(message ?? $"Input {parameterName} did not satisfy the options", parameterName);
            }

            return input;
        }
```

I'm happy to program them out for you unless there are valid arguments against it.

**Comment by ardalis** [comment]
These only make sense for custom predicates; seems like a good thing to add.

**Comment by ardalis** [comment]
Make a PR and add something here I think: https://github.com/ardalis/GuardClauses/blob/main/src/GuardClauses/GuardAgainstExpressionExtensions.cs

and I'll accept it. Thanks!

**Comment by danny-bos-developer** [comment]
I have created a pull request: https://github.com/ardalis/GuardClauses/pull/228
Could you please look into it?

**Comment by danny-bos-developer** [comment]
There was a build error that I have fixed. PR should be good now
